### PR TITLE
Internal rewrite of how Wings are defined

### LINKED ISSUE #21: Allow more complex wing geometry
Currently, all wings must have a horizontal leading edge coming out of the wing origin due to the way inertial properties of the wing are calculated. In order to do optimizations on more complex wings, this limitation must be removed. Realistically, this is going to require a refactor of how wing classes are defined in order to not make definitions of individual wings stupid. 

What I will probably end up doing is change it so that wings are made of different parts: rods, films, and curves. Rods will be cylindrical bars with a start and end point. Curves will be slender rods that follow a nonlinear path. curves will likely have a flag to either have a round profile or rectangular profile. Both of these have the math to create them already, but these will likely get their own classes somewhere. 

Films are where most of the work lies. Films are planar entities defined by two curves that share the same start and end y coordinates. Sharing the same start and end y coordinate helps keep wings more plausible. In order for the two curves to be completely arbitrary, calculations must be able to be done for curves that depend on different parametric inputs, which makes this nontrivial.

### PR DESCRIPTION
**Pull request by BrightKnight02**
The original way wings were defined was pretty restrictive an obnoxious for building a new wing to optimize. This PR contains the changes done to make the internal representation better for making different wing types.

# `Geometry` class
Fundamentally, any wing is made out of some collection of 2D shapes, all of which contribute to the overall properties of the wing. `Geometry` is a high level class that doesn't do anything at the moment, but it has many useful subclasses.
## `Curve` subclass
The basic subclass for a 1D geometry subclass. A curve is defined by four functions, each of which take an input parameter `t` between 0 and 1. The four functions define the way y, z, dy/dt, and dz/dt vary with respect to `t` for the curve in the yz plane. Curves have 3 derived properties: `length`, center of area `coa`, and second area moment `second_moment`. `length` is the total length of the curve. `coa` is a 1x2 numpy array containing the average y and z coordinates of the curve given by a pair of torque integrals divided by `length`. `second_moment` is a 1x6 numpy array containing the analogue of the moment of inertia for a curve. 

`Curve` has subclasses 'Line', "LineSegments', and `CurbicBezier` that make the creation of lines, collections of connected line segments, and cubic bezier curves respectively easier . 

## `BoundedRegion` subclass
A bounded region is defined by two curves that start and end on the same y coordinate, the upper curve having strictly greater y coordinates compared to the lower curve. The three derived properties for bounded regions are `area`, center of area `coa` and second moment of area `second_moment`. `area` is just the area of the bounded region, and the other two properties are the same as the properties of the same name for a `Curve`, just done over an area rather than along a curve. 

`BoundedRegion` has 3 subclasses: `Polygon`, `Circle`, and `Rectangle`. `Polygon` defines an n sided region with one point always on the -y axis. `Circle` creates a circle centered on the origin. `Rectangle` creates a region with one vertex in each of the quadrants of the yz plane centered on the origin. 

# `Component` class
This class defines an object with specific geometric and inertial properties. 

## `Film` subclass
Using a `BoundedRegion`, this class defines the inertial properties of a thin sheet of material with a known thickness and constant density.

# `Sweep` subclass
Takes a `Curve` path and a `BoundedRegion` profile with a constant density and calculates the inertial properties of such a sweep assuming the curve is sufficiently smooth, the curve passes through the center of area of the profile, and the path is very long compared to any dimension of the profile. 

## `Rod` subclass
Uses the same assumptions as `Sweep`, but the path is defined by coordinates in the yz plane and the profile is a circle of a given diameter

# `Wing` class
A basic constructor that takes in a `leading_edge` `Curve` and a `trailing_edge` `Curve` and a collection of `Sweep` instances `sweeps` to determine the inertial properties of a wing of the shape defined by the leading and trailing edges supported by the components in `sweeps` 

The `TriWing` class used by the current optimization was edited to use this new internal format, but it is interacted with in the same way as the old version. 

# Other Changes
Added a helper function to adjust the inertia tensor if it is barely failing triangle tests.

### CODE REVIEW

**Comment by BrightKnight02** [comment]
Technically once all the wings and tests are are migrated over to the new format, `inertial_properties.py` can be deleted

File: flapping_proj/optimization/wing_classes.py [hunk 1]
```diff
@@ -21,74 +69,47 @@ def __init__(self, y_0, z_0, y_1, z_1, y_2, z_2, y_3, z_3):
...
     
-    def dz(self, t):
+    def lower_z(self, t):
+        return (1 - t)**3 * self.z0 + 3*(1 - t)**2 * t * self.z1 + 3*(1 - t) * t**2 * self.z2 + t**3  * self.z3
+    
+    def lower_dz(self, t):
         return 3 * (1 - t)**2 * (self.z1 - self.z0) + 6 * (1 - t) * t * (self.z2 - self.z1) + 3 * t**2 * (self.z3 - self.z2)
     
-        
+    def upper_z(self, t):
```
Comment (darint916) [inline flapping_proj/optimization/wing_classes.py:84]: is this still used? questionable

File: flapping_proj/optimization/wing_classes.py [hunk 2]
```diff
@@ -1,7 +1,55 @@
...
+            del_z = i.com[1] - self.com[1]
+            displacement_radius = (del_y)**2 + (del_z)**2
+            self.I += i.I + i.mass * np.array([displacement_radius, displacement_radius - del_y**2, displacement_radius - del_z, 0, 0, -del_y * del_z])
+        
+        self.I = inertia_modifier(self.I)
+        
+        com_magnitude_sqr = self.com[0]**2 + self.com[1]**2
+        self.I_origin = self.I + self.mass * np.array([com_magnitude_sqr, com_magnitude_sqr - self.com[0]**2, com_magnitude_sqr - self.com[1], 0, 0, -self.com[0]*self.com[1]])
 
 class BezierWing():
```
Comment (darint916) [inline flapping_proj/optimization/wing_classes.py:54]: add deprecated comment if not used

File: flapping_proj/unit_tests/test_film_inertial_properties.py
Comment (darint916) [inline flapping_proj/unit_tests/test_film_inertial_properties.py:1]: note file deprecation once bezier wing updated

Comment (darint916) [reply flapping_proj/unit_tests/test_film_inertial_properties.py:1]: note needs to be updated

**Review by darint916 — APPROVED** [review]
well documented. Test cases ran and passed, but untested on end to end full test

**Pull request merged**